### ISSUE TITLE
Use workspace cargo

### ISSUE THREAD

**Issue by HKalbasi** [rust-analyzer/rust-analyzer#10445]
When I was working on clippy, rust analyzer was broken due to some error related to edition 2021. It mentions that I was using cargo 1.55 but `cargo --version` in vscode terminal returns 1.57 (in outside terminal it was 1.55). It fixed when I switched to nightly globally, but I think rust analyzer should use workspace cargo.

**Comment by lnicola** [comment]
Do you mean the one in rust-toolchain.toml?

**Comment by HKalbasi** [comment]
hmm, yes. I think there is no other way to change cargo just for one workspace.

**Comment by lnicola** [comment]
Strange, those should already work, as far as I know.

**Comment by HKalbasi** [comment]
This is the error I get if I change global toolchain to stable:

```

rust-analyzer failed to load workspace: Failed to read Cargo metadata for Rust sources: Failed to run `cargo metadata --manifest-path /home/hamid/.rustup/toolchains/nightly-2021-09-28-x86_64-unknown-linux-gnu/lib/rustlib/rustc-src/rust/compiler/rustc/Cargo.toml`: `cargo metadata` exited with an error: error: failed to parse manifest at `/home/hamid/.rustup/toolchains/nightly-2021-09-28-x86_64-unknown-linux-gnu/lib/rustlib/rustc-src/rust/compiler/rustc/Cargo.toml`

Caused by:
  feature `edition2021` is required

  The package requires the Cargo feature called `edition2021`, but that feature is not stabilized in this version of Cargo (1.55.0 (32da73ab1 2021-08-23)).
  Consider trying a newer version of Cargo (this may require the nightly release).
  See https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#edition-2021 for more information about the status of this feature.
```

Maybe it has something special to clippy?

**Comment by lnicola** [comment]
Hmm, does that happen outside of rust-lang/rust?

**Comment by HKalbasi** [comment]
Yes. Development of clippy is outside of rust repo and only maintainers update it in that.

**Comment by bjorn3** [comment]
Clippy depends on nightly, so you can't use stable. Using stable in combination with `rust-analyzer.rustcPath` is not guaranteed to work and doesn't work right now as most crates in rust-lang/rust switched to the 2021 edition, which is unstable on the current stable version of rustc/cargo.

**Comment by HKalbasi** [comment]
Clippy declared nightly in rust-toolchain.toml so RA should automatically use nightly? This is the point, if you addressed it I didn't understand.

**Comment by bjorn3** [comment]
Ah, right. Maybe rust-analyzer doesn't use the clippy workspace as working directory when running `cargo metadata` for the rustc source and as such rustup uses the global default toolchain?
